### ISSUE TITLE
Distinguish query and update calls in await

### ISSUE THREAD

**Issue by chenyan-dfinity** [dfinity/motoko#3898]
Forwarding a user request from Chinese devs:

Currently, inter-canister calls for both update and query methods follow the same syntactic form: `await update_method()` and `await query_method()`. But semantically, `query_method` rolls back the state after await. It would be nice to syntactically distinguish calling update and query methods, e.g. `await query query_method()`. So that compilers can statically check the misuse, e.g. calling a query method when it is supposed to call an update method.

**Comment by nomeata** [comment]
I'm not convinced. Whether the state is rolled back or not is mostly an implementation detail of the callee, and the caller should ideally not have to care (at least for inter-canister calls; for external calls things are a bit different, which is why we have the annotation on the public interface in the first place).

It might be useful to add warnings when the _implementation_ of a query method changes global state, though.

**Comment by rossberg** [comment]
I agree with @nomeata, this is putting the concern on the wrong side of the call.

Furthermore, `await` is completely independent from calls, since the operand can be any expression with async. So this coupling wouldn't even work conceptually, unless async types and the type system were enriched with a rudimentary form of effect system.

**Comment by iclighthouse** [comment]
Then it can be put another way. If Canister A wants to call Canister B's `query test1()` method, and if Canister B's test1() is an update method, then Canister A doesn't call it. 
How can Canister A determine that test1() is a query method?

**Comment by rossberg** [comment]
@iclighthouse, if I understand correctly, you are asking about reflection, which is an entirely different question. And to be honest, one I'm even more skeptical about, because it's usually a red flag if interfacing needs reflection. Concretely, why would you care?

**Comment by chenyan-dfinity** [comment]
I think it's about the type checker not using an available information. For example, calling `f : (Text) -> async ()` with a number is a type error. When making inter-canister calls, the type system ignores whether it's an update or query method, as long as it's an async function. @iclighthouse is asking if we can make a more fine-grained control over whether we are awaiting an query call or update call. I think it matters when we try to analyze the correctness of async programs. With query call, we can assume no state change after await. Things may become more complicated when we have composite query calls.

**Comment by nomeata** [comment]
I still don't get what we want to check about the _call_ to query/non-query method? What would be an example of “wrong” code?

**Comment by chenyan-dfinity** [comment]
I think it's only a syntactic check. Suppose there are two public endpoints `shared insert_update()` and `shared query insert_dryrun()` which both calls an internal state-mutating function `insert`. The user may want to call `insert_dryrun` but instead calls `insert_update`. If we have a syntactic form like `await_query`, the compiler can catch this bug.

I think there is not much we can check on the callee side, as many query methods do change the state.

**Comment by iclighthouse** [comment]
There is indeed such a need to facilitate developers to make innovative applications in some control details.
For example, cross-canister queries that do not update any state.
Now the query method is not possible to implement cross-canister calls. With the above `query call` feature, implementing similar functionality in the update method can at least control that no state will be changed. Specifically, I just wanted to write a `query getSomething()` in Canister A, but getSomething() has to call `query getXXX()` in Canister B. So we can only modify Canister A's function to `(update) getSomething()`, and then call B.getXXX(). But without the mandatory `query call` check, the programmer might have called `(update) B.getYYY()`, which would compile and run without error.

**Comment by nomeata** [comment]
> it needs to call another Canister method as instructed

An inter-canister call is an inter-canister call, no matter what the callee does with it (update the state or not). The proxy cannot “behave differently”.

I sense some confusion with how the distinction matters for external calls, where we indeed have to distinguish between query and update calls?

**Comment by iclighthouse** [comment]
In memory, uint and bytes8 are the same, so why make the distinction? But it's not the same for developers. I think this example illustrates the need very well.

**Comment by nomeata** [comment]
But we are talking about exposing an internal detail of the implementation one canister to the developer of another canister. If it weren't for the unfortunate complications on the HTTP interface, we wouldn't even have the distinction between query and update method in the interface!

**Comment by iclighthouse** [comment]
It may not be exactly what you understand.
To get there is easy to do, just judge and throw exceptions based on the actor's candid interface. Each method is marked query/update when it is declared, which does not involve internal details. And the candid is inherently public.

**Comment by crusso** [comment]
I'm not convinced of the necessity of this, but it seems the natural place to enforce the restriction would be on the call, not the `await`.  Written more explicitly as:

```
actor A{
    public func getInfo1() : Text{
       ........
        let a = query B.test1(); // statically ok
        await a
    };
    public func getInfo2() : Text{
        ......
        let a = query B.test2(); // statically rejected
       await a
    };
}

actor B{
    public query func test1() : Text{
        "Hello!"
    };
    public func test2() : Text{
         "Hello!"
    };
}
```
Since `query` is part of the type of a canister method, we could check this statically. But I'm not convinced we should.
